Do not remove container if underlying resource cleanup failed

rhvgoyal opened the pull request:
This should fix issue #10955 and #6746 

This patch series does two things. First of all it does not remove container from a user's view if container removal failed. Secondly, it extends force option (-f) to forcibly remove the container even if resource cleanup failed. That way those user who are willing to let resources leak, can use --force.

rhvgoyal commented:
Hmm..., there are more calls to daemon.Rm(). Will fix those to adjust for additional parameter.

rhvgoyal commented:
Pushed one more time fixing all the instances of daemon.Rm(). Hopefully I got them all.

jfrazelle commented:
can you add fixes #6746 to your first comment so when this is merged it will close that issue as well

jfrazelle commented:
Also is there anyway to add an integration test

rhvgoyal commented:
@jfrazelle 

Where should I add "fixes #6746" now? Create a new pull request?

rhvgoyal commented:
@jfrazelle 

I just edited my initial pull request comment. I guess that's what you were referring to.

I will look into current integration tests and see if they are sufficient to test this or not.

jfrazelle commented:
oh just in the first comment

On Wed, Mar 4, 2015 at 6:46 AM, Vivek Goyal <notifications@github.com>
wrote:

> @jfrazelle <https://github.com/jfrazelle>
>
> Where should I add "fixes #6746
> <https://github.com/docker/docker/issues/6746>" now? Create a new pull
> request?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/11137#issuecomment-77169770>.
>

jfrazelle commented:
editing the comment should do it, no need for a new PR

On Wed, Mar 4, 2015 at 6:47 AM, Jessica Frazelle <jess@docker.com> wrote:

> oh just in the first comment
>
> On Wed, Mar 4, 2015 at 6:46 AM, Vivek Goyal <notifications@github.com>
> wrote:
>
>> @jfrazelle <https://github.com/jfrazelle>
>>
>> Where should I add "fixes #6746
>> <https://github.com/docker/docker/issues/6746>" now? Create a new pull
>> request?
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/docker/docker/pull/11137#issuecomment-77169770>.
>>
>
>

rhvgoyal commented:
@jfrazelle 

I see that integration-cli/docker_cli_rm_test.go already has bunch of container removal tests. I think adding two more tests to it should make sense.

- Create a failure situation so that container removal fails and container state is "Dead".
- Use "docker rm -f" to be able to remove a dead container.

But I am not sure how to create that failure situation reliably. In my testing I use device mapper driver and bind mount container thing device somewhere so that device deactivation fails which in turn leads to container removal failure. But it is not generic method and I can't see an easy way to force it. 

I am looking for something which can be easily used with all graphdrivers and can cause container removal failure. Any ideas?

cpuguy83 commented:
@rhvgoyal Really you should be able to mount anything into the container fs path and then try to remove the container, this should result in `Device or resource busy` shouldn't it, since there is another fs mounted on top?

rhvgoyal commented:
@cpuguy83 

I think there are couple of issues.

- I am assuming that daemon is running in a separate mount namespace. And if I bind mount something in client namespace, that would not help.

- Even if I figure out a way to get into daemon mount namespace and bind mount something else on container rootfs, it will not work with my pending proposed patches.  I have proposed another set of patches to unmount container rootfs from daemon mount namespace after successful container start to solve various issues. Once that is in, that means,  container rootfs has already been unmounted and once container exits only device will be deactivated. So bind mounting something on container rootfs will not help.

I proposed my patches here.

https://github.com/docker/docker/pull/11114

rhvgoyal commented:
@jfrazelle 

So generic container removal tests are already there. What is missing is couple of tests for testing behavior in case of failure. And right now I can't find a good way to force failure.

Can these patches be merged without tests.

rhatdan commented:
@jfrazelle @crosbymichael @cpuguy83 We would like to get this in for docker-1.6 if possible.

jfrazelle commented:
@rhatdan I added the milestone

LK4D4 commented on daemon/delete.go:
```diff
@@ -61,8 +61,13 @@ func (daemon *Daemon) ContainerRm(job *engine.Job) engine.Status {
+				log.Errorf("Cannot destroy container %s: %s", name, err)
+			} else {
+
+				return job.Errorf("Cannot destroy container %s: %s", name, err)
```
This condition can be reversed for readability:
```
if !forceRemove {
    return job.Errorf("Cannot destroy container %s: %s", name, err)
}
log.Errorf("Cannot destroy container %s: %s", name, err)
```

rhvgoyal replied:
Ok, Will change. This does look more readable.

LK4D4 commented on daemon/delete.go:
```diff
@@ -92,13 +97,29 @@ func (daemon *Daemon) Rm(container *Container) error {
-	daemon.idIndex.Delete(container.ID)
-	daemon.containers.Delete(container.ID)
+	// Mark container dead. We don't want anybody to be restarting it.
+	container.SetDead()
```
Shouldn't you check error here?

rhvgoyal replied:
As you noted in next comment, SetDead() does not return error.

LK4D4 commented on daemon/state.go:
```diff
@@ -217,3 +232,26 @@ func (s *State) IsPaused() bool {
+	s.Unlock()
+}
+
+func (s *State) SetDead() error {
```
Ah, so, there is no returning error.

cpuguy83 replied:
If this is not returning an error, can we remove `error` here?

rhvgoyal commented:
Pushed top most patch again to take care of comment from @LK4D4

cpuguy83 commented on daemon/container.go:
```diff
@@ -326,6 +326,14 @@ func (container *Container) Start() (err error) {
 	}
 
+	if container.RemovalInProgress {
+		return fmt.Errorf("Container is already being removed. Can not start it.")
```
Can remove `already`, which is confusing because the user is calling `start`

rhvgoyal replied:
Makes sense. Will do.

cpuguy83 commented on daemon/container.go:
```diff
@@ -326,6 +326,14 @@ func (container *Container) Start() (err error) {
+	}
+
+	if container.Dead {
+		return fmt.Errorf("Container is already dead. Can not start it.")
```
Can remove already, which is confusing because the user is calling start
Also, what does `dead` mean?

rhvgoyal replied:
Dead means that container can't be started now. One can only try to remove it again. (After previous removal attempt had failed). A dead container might have partially cleaned up its resources already and only some of the resources still might be there.

For example, If I bind mount a container device somewhere, then container removal will fail. It will be in "Removal In Progress" state for 10 seconds and then transition into "Dead" state. 

A user can try to figure out what was wrong, why container removal failed. Fix that condition and try removal again. 

Had to keep a separate state to mark that container is partially destroyed otherwise one might want to start it.

rhatdan replied:
Would zombie clear it up  :^)

rhvgoyal replied:
@rhatdan 

Zombie sounds good too. I will use whatever people prefer. I personally am little inclined towards "Dead".

cpuguy83 commented on daemon/delete.go:
```diff
@@ -92,13 +95,29 @@ func (daemon *Daemon) Rm(container *Container) error {
+	// If force removal is required, delete container from various
+	// indexes even if removal failed.
+	defer func() {
+		if err != nil && forceRemove {
```
I think this `err` is not in scope of any of the bellow `err`'s (or any of the `err's in the whole function), so would always be nil.

rhvgoyal replied:
I tested it and err is not nil.

I have defined a named return error as follows.

func (daemon *Daemon) Rm(container *Container, forceRemove bool) (err error)

So this err inside defer must be referring to that. And while I am returning, I am always explicitly returning an error object. So "err" ends up referring to that. 

But if it is confusing and if you have a preferred way to handle it I can do that. I guess I can stop redeclaring err in various calls and that might help.

cpuguy83 replied:
But the err in defer is nil, see example:
http://play.golang.org/p/rPGaYb3bVA

Unless I'm missing something (which is entirely possible).

rhvgoyal replied:
@cpuguy83 

Argument to defer are calculated at the time of defer and that's why you are seeing err nil. In my case, I don't pass err as argument to deferred function.

rhvgoyal replied:
For example, I modified your code a bit to not pass err as an argument to deferred function.

https://play.golang.org/p/r9nxzUbbT4

cpuguy83 replied:
Oh, of course! Thanks for taking the time to clarify, sorry for the noise here!

tiborvass replied:
@rhvgoyal shouldn't the condition be `if err == nil || forceRm` ? that way you don't need to copy this part at the end

cpuguy83 commented on daemon/container.go:
```diff
@@ -326,6 +326,14 @@ func (container *Container) Start() (err error) {
 		return nil
 	}
 
+	if container.RemovalInProgress {
```
We can probably condense these two errors into one because in terms of the end user, once a container is marked for removal at all it can't be used anyway (regardless if it got to `dead` state or not).

So:
```
if container.RemovalInProgress || container.Dead {
    return fmt.Errorf("container is marked for removal and cannot be started")
}
```

rhvgoyal replied:
Ok, Will do.

cpuguy83 commented on daemon/delete.go:
```diff
@@ -61,8 +61,11 @@ func (daemon *Daemon) ContainerRm(job *engine.Job) engine.Status {
+			if !forceRemove {
+				return job.Errorf("Cannot destroy container %s: %s", name, err)
+			}
+			log.Errorf("Cannot destroy container %s: %s", name, err)
```
Should be `"Cannot destroy container %s: %v"

rhvgoyal replied:
Will change.

rhvgoyal commented:
Something interesting is happening. The container I removed using -f, comes back to life if I restart the docker service. Wow, I need to look into it. May be we are not deleting underlying files somewhere and restart scans for these files and gets it back.

rhvgoyal commented:
Ok, /var/lib/docker/container/<container_id> directory is removed after container root and container init root has been removed. 

So it depends on when failure occurs. If failure occurs before container file removal, it will come back to life. And I think it is a bad idea. Container might be half deleted and if somebody tries to start it, bad things will happen. I will save the container state "Dead" back to disk before proceeding with actual deletion. This will make sure that once container is marked dead, it remains dead even if docker is restarted.

rhvgoyal committed:
container: Do not remove contianer if any of the resource failed cleanup

Do not remove container if any of the resource could not be cleaned up. We
don't want to leak resources.

Two new states have been created. RemovalInProgress and Dead. Once container
is Dead, it can not be started/restarted. Dead container signifies the
container where we tried to remove it but removal failed. User now needs to
figure out what went wrong, corrent the situation and try cleanup again.

RemovalInProgress signifies that container is already being removed. Only
one removal can be in progress.

Also, do not allow start of a container if it is already dead or removal is
in progress.

Also extend existing force option (-f) to docker rm to not return an error
and remove container from user view even if resource cleanup failed.
This will allow a user to get back to old behavior where resources
might leak but atleast user will be able to make progress.

Signed-off-by: Vivek Goyal <vgoyal@redhat.com>

rhvgoyal commented:
Ok, pushed one more time. Have taken care of the review comments so far. And also took care of making "Dead" state persistent to disk.

rhvgoyal commented:
@jfrazelle 

This windows test failure seems unrelated to my patches.

jfrazelle commented:
I reran and it passes, sorry we are working through some windows CI things

estesp commented:
So this is shaping up to be a slightly different implementation of my PR (#10099) from awhile ago; although there are some better choices here about ordering of operations (rather than remove and re-add).  The naming piece is still difficult, but I'm fine with "Dead" if that is reasonable for others.

The one piece I don't like is that I was preserving the failing error--whatever kept the container from being removed successfully is now a transient error that after hours (or days or weeks?) is now lost, although I guess another try at `docker rm` will resurface it?  However, having it available for `docker ps` was something I liked about my impl, although it added the painful UI problem of "way too long/wide output".

rhvgoyal commented:
@estesp 

I think storing the error in container is not a bad idea. And it is not orthogonal to my patches. Right now I am trying to fix the problem of not removing container from main view if deletion failed and create new container state to represent that.

If this works well, then next step can be to also store the failure message. And there can be a separate patch for that. But I don't think that storing error should be a requirement for this patch series to go in.

jfrazelle commented:
look like needs a rebase sorry :(

LK4D4 commented on daemon/state.go:
```diff
@@ -217,3 +232,26 @@ func (s *State) IsPaused() bool {
+	s.Unlock()
+}
+
+func (s *State) SetDead() error {
```
It can't return non-nil error, why have confusing signature?

rhvgoyal replied:
Makes sense. Will remove it.

LK4D4 commented on daemon/state.go:
```diff
@@ -42,6 +44,14 @@ func (s *State) String() string {
+		return "Removal In Progress"
+	}
+
+	if s.Dead {
```
I'm not totally sure, but you checking for `s.Dead` twice.

rhvgoyal replied:
Hm...., I don't understand this comment. I am first checking for Removal In Progress and dispalying that. Yes a container could be dead as well as removal is in progress. In that case I think it makes sense to display "Removal In Progress".

If that's not set, then I check for "Dead" state and display "dead". So where is the redundant check?

LK4D4 replied:
line 74 I mean

rhvgoyal replied:
But that's a separate call in separate function. First call is in String() and second call is in StateString().  So does not look like there is redundancy.

LK4D4 replied:
Lol, indeed. Sorry :blush:

LK4D4 commented on daemon/state.go:
```diff
@@ -11,16 +11,18 @@ import (
+	Paused            bool
+	Restarting        bool
+	OOMKilled         bool
+	removalInProgress bool // Not need for this to be persistent on disk.
```
could you move this below? to have sorta separation between public and private

rhvgoyal replied:
Ok, will do.

rhvgoyal replied:
@LK4D4 

Now I am thinking that it is better to keep it where it is. Reason being that right now all the possible container states are together. If I move it below waitChan, then a single state is out of the group. And I think it is better to keep all states together instead of forcing grouping based on public/private.

LK4D4 commented on daemon/delete.go:
```diff
@@ -92,19 +95,40 @@ func (daemon *Daemon) Rm(container *Container) error {
+
+	defer container.ResetRemovalInProgress()
+
+	if err = container.Stop(3); err != nil {
```
From where `3` number? Shouldn't be that constant?

rhvgoyal replied:
That's existing code. I did not introduce it. I agree that using a constant is much better. But that cleanup is probably should be part of a separate patch series.

LK4D4 replied:
Ok, agree

LK4D4 commented on daemon/delete.go:
```diff
@@ -92,19 +95,40 @@ func (daemon *Daemon) Rm(container *Container) error {
 		log.Debugf("Unable to remove container from link graph: %s", err)
 	}
 
-	if err := daemon.driver.Remove(container.ID); err != nil {
```
I don't understand why you need to remove `:`. It is enough if returned value override top-level error.

rhvgoyal replied:
Hmm..., What's the advantage of retaining ":"?. Only case where we don't want to override err is when we don't want to return actual error. Something like as follows. And there are easy ways to override that in return statement. 

if  err := foo(); err != nil {
     // This is not catastropic. Return nil.
     return
}

Instead following is easier to read because it does not rely on the previous value of err.

if err = foo(); err != nil {
      // This is not a catastropic error
       return nil
}

So not using ":" looks fine to me in above code example.

LK4D4 replied:
Just to reduce patch size. I'm not insisting if you like it.

LK4D4 commented on daemon/delete.go:
```diff
@@ -82,7 +85,7 @@ func (daemon *Daemon) DeleteVolumes(volumeIDs map[string]struct{}) {
 
 // Destroy unregisters a container from the daemon and cleanly removes its contents from the filesystem.
-func (daemon *Daemon) Rm(container *Container) error {
+func (daemon *Daemon) Rm(container *Container, forceRemove bool) (err error) {
```
Okay, I'm not in favor of this signature. How about helper function and two separate functions `Rm` and `ForceRm` which will call helper with `true` or `false`?

rhvgoyal replied:
Ok, I can do that. That will reduce the size of changes as all the existing calls to Rm() will not have to be modified.

rhvgoyal commented:
@jfrazelle 

Why are we removing this from 1.6.0 milestone?

jfrazelle commented:
Since its a bug fix we can include it at anytime in an RC but I don't want
it to hold up RC1 unless you can make these changes in the next hour :)

On Monday, March 23, 2015, Vivek Goyal <notifications@github.com> wrote:

> @jfrazelle <https://github.com/jfrazelle>
>
> Why are we removing this from 1.6.0 milestone?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/11137#issuecomment-85168695>.
>

rhvgoyal commented:
@jfrazelle 

I can do it but I don't like to rush in the changes as it always increases the possibility of breaking something.

All the comments are minor code cleanup commands, nothing major. It does not sound fair that just before RC1, one can give code cleanup comments and block the inclusion of fixes. 

Can't we take these in 1.6.0 and I can open a separate issue to follow up on code cleanup comments.

jfrazelle commented:
That's sounds reasonable to me

On Monday, March 23, 2015, Vivek Goyal <notifications@github.com> wrote:

> @jfrazelle <https://github.com/jfrazelle>
>
> I can do it but I don't like to rush in the changes as it always increases
> the possibility of breaking something.
>
> All the comments are minor code cleanup commands, nothing major. It does
> not sound fair that just before RC1, one can give code cleanup comments and
> block the inclusion of fixes.
>
> Can't we take these in 1.6.0 and I can open a separate issue to follow up
> on code cleanup comments.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/11137#issuecomment-85169616>.
>

rhvgoyal commented:
@LK4D4 

Would you have an objection to including these fixes as it is in 1.6.0 and I follow up with your cleanup comments in a separate issue for next release?

LK4D4 commented:
@rhvgoyal Dunno, do you want me to prepare PR for your branch. Because I don't want to have changed signature in release. This can lead to some nasty merges.

jfrazelle commented:
we can always include it in another RC when it is ready

jfrazelle commented:
I would rather wait and have it be good

rhvgoyal commented:
@jfrazelle @LK4D4 

Ok, I will make changes and generate another PR. Hopefully we can pull it in next RC release.

jfrazelle commented:
well just push to this one :) we can then include all of it in one RC

rhvgoyal commented:
@jfrazelle 

Ok.

rhvgoyal commented:
@jfrazelle @LK4D4 

Pushed changes again after taking care of some of the comments. Please have another look.

And this build failure seems unrelated to my changes.

LK4D4 commented:
I don't see any big problems
LGTM

estesp commented:
This looks ready, but I still think it makes sense to test this functionality as it has new codepaths that current tests won't exercise.  Given our `integration-cli` environment is well-known, and many tests assume client == daemon/host, then wouldn't a simple file open inside the container's mount dir disallow unmounting, causing a "busy" error and removal to fail (in the same way that an extra mount would, but much more simply).  The path to the mount point should be easy to discern, given our knowlege of the test harness/daemon config, right?

rhvgoyal commented:
@estesp 

- How do we know the mount point of container? Docker daemon does  not export that and hardcoding that in tests is not a good idea as if we change it later, test will be broken.

- Mount point will also depend on graphdriver being used. I am not sure how to query daemon for graphdriver in use.

- I did not understand your comment about client == daemon/host. Can you elaborate  bit more. And also point me to a test which assumes it.

rhvgoyal commented:
@estesp 

Also I think it is not a good idea to assume that container root is mounted in daemon namespace after container launch. I have another PR request waiting to change that behavior.

https://github.com/docker/docker/pull/11114

I think to make it work, we will have to have some sort of mount/unmount commands from docker so that it can mount a container's root for us explicitly and allow us to query mount point of container explicitly. That will work only if daemon is not running in a separate mount namespace.

jfrazelle commented:
needs a rebase also can you squash to 1 commit otherwise cherry-picking into the release branch is going to be kinda annoying

rhvgoyal commented:
@jfrazelle 

rebased and put everything into a single patch and pushed one more time. Tests are passing on my local system. So there might be something wrong with the build.

cpuguy83 commented:
Testing this manually and it works.
Looks like CI is having an issue pulling lxc :(

jfrazelle commented:
If it fails lxc we aren't merging can you rebuild

On Tuesday, March 31, 2015, Brian Goff <notifications@github.com> wrote:

> Testing this manually and it works.
> Looks like CI is having an issue pulling lxc :(
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/11137#issuecomment-88162127>.
>

vbatts commented:
linuxcontainers.org is down :-(

jfrazelle commented:
we will wait...

vbatts commented:
```
--- FAIL: TestEventsFilterContainerID (0.94s)
	docker_cli_events_test.go:355: Expected 3 events, got 2: [2015-03-31T17:36:02.000000000Z ab26eddbd0f26fad7fb1263785e701e23d5f5021afbb196a94d4c31c06a97de9: (from busybox:latest) create 2015-03-31T17:36:02.000000000Z ab26eddbd0f26fad7fb1263785e701e23d5f5021afbb196a94d4c31c06a97de9: (from busybox:latest) start]
```

jfrazelle commented:
ya that one has been racy lately

On Tue, Mar 31, 2015 at 10:51 AM, Vincent Batts <notifications@github.com>
wrote:

> --- FAIL: TestEventsFilterContainerID (0.94s)
>     docker_cli_events_test.go:355: Expected 3 events, got 2: [2015-03-31T17:36:02.000000000Z ab26eddbd0f26fad7fb1263785e701e23d5f5021afbb196a94d4c31c06a97de9: (from busybox:latest) create 2015-03-31T17:36:02.000000000Z ab26eddbd0f26fad7fb1263785e701e23d5f5021afbb196a94d4c31c06a97de9: (from busybox:latest) start]
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/11137#issuecomment-88186867>.
>

jfrazelle commented:
@estesp what are your thoughts

jfrazelle commented:
are you ok with this getting in 1.6.0, it needs to be merged today otherwise, since it is not really a regression and was a bug in earlier versions I will remove from the milestone

jfrazelle commented:
passes all tests @vbatts you think ok to merge?

vbatts commented:
so after learning up on this PR, it will make clean up of containers much nicer.
Like happens occasionally where a container fails to be removed (outside mounts, mount namespace overlap, etc), it will leave the container and underlying state to where it needs to be cleaned up by hand.

With this PR, now it is possible to get the error on `docker rm ...` earlier, fix the issue, then `docker rm ...` again. Without having to bring down the daemon for container cleanup.

I've just walked through a couple of cases by hand, and it's improved the experience.

LGTM

tiborvass commented on daemon/delete.go:
```diff
@@ -94,19 +109,40 @@ func (daemon *Daemon) Rm(container *Container) error {
 		logrus.Debugf("Unable to remove container from link graph: %s", err)
 	}
 
-	if err := daemon.driver.Remove(container.ID); err != nil {
```
@rhvgoyal why remove the `:` ?

tiborvass replied:
oh for the defer...

tiborvass commented on daemon/delete.go:
```diff
@@ -83,8 +90,16 @@ func (daemon *Daemon) DeleteVolumes(volumeIDs map[string]struct{}) {
+
 // Destroy unregisters a container from the daemon and cleanly removes its contents from the filesystem.
-func (daemon *Daemon) Rm(container *Container) error {
+func (daemon *Daemon) commonRm(container *Container, forceRemove bool) (err error) {
```
I'd much rather keep `Rm` as the name, and just add the bool as a param. That way the calling code above can be reduced to: `if err := daemon.Rm(container, forceRemove); err != nil {`

rhvgoyal replied:
@tiborvass 

That's how it was initially and later @LK4D4 preferred it this way. So I changed it.

rhvgoyal replied:
For the record, I liked passing force as a parameter to Rm() better instead of creating another variant. It ended up in more lines of changes as every existing call to Rm() had to be modified but I still had liked it better.

tiborvass replied:
@rhvgoyal huh I see, I'll try to convince @LK4D4 in another PR :P

jfrazelle commented:
ah shoot sorry @tiborvass my GH page hadn't refreshed until after the button click!
